### PR TITLE
Integrated group/cluster lenses based on catalogs in the LensPop interface

### PR DESCRIPTION
**Pull request by furcelay**
Implements a new class `ClusterCatalogLens` that stores a catalog of lenses based on a group or cluster catalog. This requires two astropy tables: one for the cluster properties, and one for the members, along with a list of galaxies from the SkyPy pipeline. 

This class assigns a NFW halo to each cluster based on the richness, and a EPL subhalo to each cluster member by choosing a  SkyPy galaxy with similar magnitudes and redshift.

I've also updated the `cluster_lens.ipynb` notebook to include an example of usage.

### CODE REVIEW

File: slsim/Deflectors/cluster_catalog_lens.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+            Mandatory keys: 'cluster_id' 'z', 'richness'
+        :param members_list: list of dictionary with lens parameters of
+            elliptical galaxies from a group/cluster member catalog.
+            Mandatory keys: 'cluster_id', 'ra', 'dec', 'mag_{band}'
+        :param kwargs_cut: cuts in parameters: band, band_mag, z_min, z_max
+        :type kwargs_cut: dict
+        :param kwargs_mass2light: mass-to-light relation
+        :param cosmo: astropy.cosmology instance
+        :type sky_area: `~astropy.units.Quantity`
+        :param sky_area: Sky area over which galaxies are sampled. Must be in units of
```
Comment (nkhadka21) [inline slsim/Deflectors/cluster_catalog_lens.py:52]: For sky_area, we use astropy Quantity(). Is this astropy Quantity?

File: slsim/Deflectors/cluster_catalog_lens.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+        )
+        self.deflector_profile = "NFW_CLUSTER"
+        self.richness_fn = richness_fn
+        self.set_cosmo()
+
+        # cluster
+        n_clusters = len(cluster_list)
+        cluster_column_names = cluster_list.columns
+        if "richness" not in cluster_column_names:
+            raise ValueError("richness is mandatory in cluster catalog")
```
Comment (nkhadka21) [inline slsim/Deflectors/cluster_catalog_lens.py:72]: This line is not tested.

File: slsim/Deflectors/cluster_catalog_lens.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        if "z" not in cluster_column_names:
+            raise ValueError("redshift is mandatory in cluster catalog")
```
Comment (nkhadka21) [inline slsim/Deflectors/cluster_catalog_lens.py:74]: This line is also not tested.

File: slsim/Deflectors/cluster_catalog_lens.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        :return: dictionary of complete parameterization of deflector
+        """
+        index = random.randint(0, self._num_select - 1)
+        deflector = self.draw_cluster(index)
+        members = self.draw_members(deflector["cluster_id"])
+        deflector["subhalos"] = members
+        return deflector
+
+    def draw_cluster(self, index):
+        cluster = self._cluster_select[index]
```
Comment (nkhadka21) [inline slsim/Deflectors/cluster_catalog_lens.py:162]: Missing documentation for the input parameter.

File: slsim/Deflectors/cluster_catalog_lens.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+        if cluster["e1_mass"] == -1 or cluster["e2_mass"] == -1:
+            e, phi = gene_e_ang_halo(np.array([cluster["halo_mass"]]))
+            e1, e2 = phi_q2_ellipticity(np.deg2rad(phi[0]), 1 - e[0])
+            cluster["e1_mass"] = e1
+            cluster["e2_mass"] = e2
+        return dict(cluster)
+
+    def draw_members(
+        self, cluster_id, center_scatter=0.2, max_dist_arcsec=80, bcg_band="r"
+    ):
```
Comment (nkhadka21) [inline slsim/Deflectors/cluster_catalog_lens.py:184]: Missing documentation of input parameters.

File: slsim/Deflectors/richness2mass.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import numpy as np
+
```
Comment (nkhadka21) [inline slsim/Deflectors/richness2mass.py:2]: Maybe write one line description for this module.

File: slsim/Deflectors/richness2mass.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+def general_mass_richness(rich, alpha, rich_0, m_0, ln_m_scatter, min_mass=0.0):
+    """M(N) = M_0 * (N / N_0) ** alpha VAR[ln(M)] = alpha**2 / N + lnM_scatter**2."""
```
Comment (nkhadka21) [inline slsim/Deflectors/richness2mass.py:5]: Missing documents for the input parameters.

File: slsim/Deflectors/richness2mass.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    m_var = ln_m_var * m**2  # VAR[log(X)] ~ VAR[X] / E[X]^2
+    m += np.random.normal(0, np.sqrt(m_var))
+    m = np.maximum(m, min_mass)
+    return m
+
+
+def mass_richness_simet2017(rich, min_mass=1e11):
+    """Mass-richness relation from Simet et al.
+
+    2017
```
Comment (nkhadka21) [inline slsim/Deflectors/richness2mass.py:18]: Missing doc for input parameters.

File: slsim/Deflectors/richness2mass.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        m_0=10**14.344,
+        ln_m_scatter=0.25,
+        min_mass=min_mass,
+    )
+
+
+def mass_richness_abdullah2022(rich, min_mass=1e11):
+    """Mass-richness relation from Abdullah et al.
+
+    2022 (sred13)
```
Comment (nkhadka21) [inline slsim/Deflectors/richness2mass.py:33]: Missing doc for the input parameters.

File: slsim/Deflectors/richness2mass.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    )
+
+
+def mass_richness_relation(rich, relation="Abdullah2022"):
+    if relation == "simet2017":
+        return mass_richness_simet2017(rich)
+    elif relation == "Abdullah2022":
+        return mass_richness_abdullah2022(rich)
+    else:
+        raise ValueError(f"Unknown mass-richness relation: {relation}")
```
Comment (nkhadka21) [inline slsim/Deflectors/richness2mass.py:46]: Please test this line.

File: slsim/Deflectors/cluster_catalog_lens.py [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+        :type cosmo: astropy.cosmology
+        :param bands: list of bands to compare
+        :type bands: list
+        :param max_gals: maximum number of galaxies to compare to
+        :type max_gals: int
+        """
+        # shuffle galaxy list and select a subset
+        if len(galaxy_list) > max_gals:
+            indices = np.random.choice(len(galaxy_list), max_gals, replace=False)
+            galaxy_list = galaxy_list[indices]
```
Comment (nkhadka21) [inline slsim/Deflectors/cluster_catalog_lens.py:244]: Is it possible to test this line?

File: slsim/Deflectors/cluster_catalog_lens.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+        mag_cols = [f"mag_{b}" for b in bands if f"mag_{b}" in members_list.columns]
+        if not bands:
+            raise ValueError("No magnitude columns found in members_list")
```
Comment (nkhadka21) [inline slsim/Deflectors/cluster_catalog_lens.py:248]: is it possible to test this line?

File: slsim/Deflectors/cluster_catalog_lens.py [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+            H0=self.cosmo.H0.value,
+            Om0=self.cosmo.Om0,
+            Ode0=self.cosmo.Ode0,
+            Ob0=self.cosmo.Ob0 if self.cosmo.Ob0 is not None else 0.04897,
+            Tcmb0=self.cosmo.Tcmb0.value if self.cosmo.Tcmb0.value > 0 else 2.7255,
+            Neff=self.cosmo.Neff,
+            sigma8=0.8102,
+            ns=0.9660499,
+        )
+        colossus_cosmo.setCosmology(cosmo_name="halo_cosmo", **params)
```
Comment (nkhadka21) [inline slsim/Deflectors/cluster_catalog_lens.py:281]: Just curious, is it necessary to use colossus cosmology or can be replaced by astropy?

File: tests/test_Deflectors/test_cluster_catalog_lens.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    galaxy_pop = cluster_catalog_lens
+    num_deflectors = galaxy_pop.deflector_number()
+    assert num_deflectors >= 0
+
+
+def test_draw_deflector(cluster_catalog_lens):
+    galaxy_pop = cluster_catalog_lens
+    deflector = galaxy_pop.draw_deflector()
+    assert deflector["z"] != 0
+    assert deflector["halo_mass"] > 0
```
Comment (nkhadka21) [inline tests/test_Deflectors/test_cluster_catalog_lens.py:55]: These test look like trivial. Can you put more specific test (maybe range of the halo mass)?

File: tests/test_Deflectors/test_cluster_catalog_lens.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    assert deflector["concentration"] > 0
```
Comment (nkhadka21) [inline tests/test_Deflectors/test_cluster_catalog_lens.py:56]: Here also, can you put specific test?

File: tests/test_Deflectors/test_cluster_catalog_lens.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    assert len(deflector["subhalos"]) > 0
```
Comment (nkhadka21) [inline tests/test_Deflectors/test_cluster_catalog_lens.py:57]: Would be great to have more specific number.

File: tests/test_Deflectors/test_richness2mass.py [hunk 17]
```diff
@@ -0,0 +1,9 @@
+from slsim.Deflectors.richness2mass import mass_richness_relation
+
+
+def test_mass_richness_relation():
+    mass = mass_richness_relation(40, relation="simet2017")
+    assert mass > 0
+
+    mass = mass_richness_relation(40, relation="Abdullah2022")
+    assert mass > 0
```
Comment (nkhadka21) [inline tests/test_Deflectors/test_richness2mass.py:9]: Here, you can compute mass using that relation and compare this number with this mass here.Then, If someone later comes and changes this relation, this test will fail.

File: tests/test_lens_pop.py [hunk 18]
```diff
@@ -145,6 +146,59 @@ def test_galaxies_lens_pop_halo_model_instance():
...
+    )
+
+    cluster_lens_pop = LensPop(
+        deflector_population=lens_clusters,
+        source_population=source_galaxies,
+        cosmo=cosmo,
+    )
+
+    kwargs_lens_cut = {}
+    pes_lens_class = cluster_lens_pop.select_lens_at_random(**kwargs_lens_cut)
```
Comment (nkhadka21) [inline tests/test_lens_pop.py:198]: Any other specific test that you want to put here?

File: slsim/Deflectors/cluster_catalog_lens.py [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+)
+from slsim.Deflectors.velocity_dispersion import vel_disp_nfw
+from slsim.Deflectors.deflectors_base import DeflectorsBase
+from lenstronomy.Util.param_util import phi_q2_ellipticity
+from astropy import units as u
+from astropy.table import hstack
+from scipy.spatial.distance import cdist
+
+
+class ClusterCatalogLens(DeflectorsBase):
```
Comment (nkhadka21) [inline slsim/Deflectors/cluster_catalog_lens.py:20]: Do you want to change this name to ClusterDeflectors? If you prefer ClusterCatalogLens, I am ok with it.

**Review by nkhadka21 — CHANGES_REQUESTED** [review]
Hi @furcelay, thank you very much for this wonderful work! It look very nice. I have some comments. Please look at them.

File: slsim/Deflectors/cluster_catalog_lens.py (hunk 1 above)
Comment (furcelay) [reply slsim/Deflectors/cluster_catalog_lens.py]: Yes, it is

File: slsim/Deflectors/cluster_catalog_lens.py (hunk 2 above)
Comment (furcelay) [reply slsim/Deflectors/cluster_catalog_lens.py]: added a  test for this

File: slsim/Deflectors/cluster_catalog_lens.py (hunk 3 above)
Comment (furcelay) [reply slsim/Deflectors/cluster_catalog_lens.py]: added a  test for this

File: slsim/Deflectors/cluster_catalog_lens.py (hunk 4 above)
Comment (furcelay) [reply slsim/Deflectors/cluster_catalog_lens.py]: added documentation

File: slsim/Deflectors/cluster_catalog_lens.py (hunk 5 above)
Comment (furcelay) [reply slsim/Deflectors/cluster_catalog_lens.py]: added documentation

File: slsim/Deflectors/richness2mass.py (hunk 6 above)
Comment (furcelay) [reply slsim/Deflectors/richness2mass.py:2]: added documentation

File: slsim/Deflectors/richness2mass.py (hunk 7 above)
Comment (furcelay) [reply slsim/Deflectors/richness2mass.py:5]: added documentation

File: slsim/Deflectors/richness2mass.py (hunk 8 above)
Comment (furcelay) [reply slsim/Deflectors/richness2mass.py:37]: added documentation

File: slsim/Deflectors/richness2mass.py (hunk 9 above)
Comment (furcelay) [reply slsim/Deflectors/richness2mass.py:56]: added documentation

File: slsim/Deflectors/richness2mass.py (hunk 10 above)
Comment (furcelay) [reply slsim/Deflectors/richness2mass.py:75]: added test

File: tests/test_Deflectors/test_cluster_catalog_lens.py (hunk 14 above)
Comment (furcelay) [reply tests/test_Deflectors/test_cluster_catalog_lens.py]: updated the test to an specific one

File: tests/test_Deflectors/test_cluster_catalog_lens.py (hunk 15 above)
Comment (furcelay) [reply tests/test_Deflectors/test_cluster_catalog_lens.py]: updated the test to an specific one

File: tests/test_Deflectors/test_cluster_catalog_lens.py (hunk 16 above)
Comment (furcelay) [reply tests/test_Deflectors/test_cluster_catalog_lens.py]: updated the test to an specific one

File: tests/test_Deflectors/test_richness2mass.py (hunk 17 above)
Comment (furcelay) [reply tests/test_Deflectors/test_richness2mass.py]: updated the test to an specific one

File: slsim/Deflectors/cluster_catalog_lens.py (hunk 19 above)
Comment (furcelay) [reply slsim/Deflectors/cluster_catalog_lens.py]: Yes, ClusterDeflectors is much better

File: slsim/Deflectors/cluster_catalog_lens.py (hunk 12 above)
Comment (furcelay) [reply slsim/Deflectors/cluster_catalog_lens.py]: added test

File: slsim/Deflectors/cluster_catalog_lens.py (hunk 11 above)
Comment (furcelay) [reply slsim/Deflectors/cluster_catalog_lens.py]: added test

**Review by nkhadka21 — APPROVED** [review]
Hi @furcelay , thank you very much for this quick update on this PR! It looks very good to me and I wait for @sibirrer 's review.

File: slsim/Deflectors/DeflectorTypes/nfw_cluster.py [hunk 20]
```diff
@@ -14,18 +15,18 @@ class NFWCluster(DeflectorBase):
     - 'e1_mass': eccentricity of NFW profile
     - 'e2_mass': eccentricity of NFW profile
     - 'z': redshift of deflector
+    - 'subhalos': list of dictionary with subhalo parameters
```
Comment (sibirrer) [inline slsim/Deflectors/DeflectorTypes/nfw_cluster.py:18]: here I would state for what class (EPLSersic for the time being)

File: slsim/Deflectors/DeflectorTypes/nfw_cluster.py [hunk 21]
```diff
@@ -14,18 +15,18 @@ class NFWCluster(DeflectorBase):
...
+    def __init__(self, deflector_dict):
         """
 
         :param deflector_dict:  parameters of the cluster halo
         :type deflector_dict: dict
-        :param subhalos_list: list with Deflector instances as cluster subhalos
-        :type subhalos_list: list[Deflector]
         """
+        subhalos_list = deflector_dict["subhalos"]
+        self._subhalos = [EPLSersic(subhalo_dict) for subhalo_dict in subhalos_list]
```
Comment (sibirrer) [inline slsim/Deflectors/DeflectorTypes/nfw_cluster.py:28]: this is ok for now, just as a comment: one might want to make optionally different deflector classes possible in future

**Review by sibirrer — APPROVED** [review]
Looks good. I had one comment and one suggestion, but happy to merge

**Comment by furcelay** [comment]
Thank you @nkhadka21  and @sibirrer for your review! It should be ready to merge now.

**Pull request merged**